Sources/SwiftWarningControl/WarningControlRegions.swift

```diff
@@ -0,0 +1,329 @@
+  public var debugDescription: String {
+    var result = "Warning Group Control Region Tree:\n"
+    func printNode(_ node: WarningControlRegionNode, indent: Int) {
+      let spacing = String(repeating: "  ", count: indent)
+      result += "\(spacing)[\(node.range.lowerBound), \(node.range.upperBound)]"
+      if !node.warningGroupControls.isEmpty {
+        result += " id(s): \(node.warningGroupControls.keys.map { $0.value }.joined(separator: ", "))\n"
+      } else {
+        result += "\n"
+      }
+      for child in node.children {
+        printNode(child, indent: indent + 1)
+      }
+    }
+
+    printNode(rootRegionNode, indent: 0)
+    return result
+  }
+}
+
+extension WarningControlRegionTree {
+  /// Determine the warning group behavior at a specified position
+  /// for a given diagnostic group
+  @_spi(ExperimentalLanguageFeatures)
+  public func warningGroupBehavior(
+    at position: AbsolutePosition,
+    for diagnosticGroupIdentifier: DiagnosticGroupIdentifier
+  ) -> WarningGroupBehavior? {
+    return rootRegionNode.innermostContainingRegion(at: position, for: diagnosticGroupIdentifier)?.behavior
+  }
+}
+
+/// A node in the warning control region tree, representing a collection of warning
+/// group controls and references to its nested child regions.
+private class WarningControlRegionNode {
+  let range: Range<AbsolutePosition>
+  var warningGroupControls: [DiagnosticGroupIdentifier: WarningGroupBehavior] = [:]
+  var children: [WarningControlRegionNode] = []
+
+  init(
+    range: Range<AbsolutePosition>,
+    for diagnosticGroupIdentifier: DiagnosticGroupIdentifier,
+    behavior: WarningGroupBehavior
+  ) {
+    self.range = range
+    self.warningGroupControls = [diagnosticGroupIdentifier: behavior]
+  }
+
+  init(range: Range<AbsolutePosition>) {
+    self.range = range
+    self.warningGroupControls = [:]
+  }
+
+  /// Check if this node's bounds contain a position
+  /// The regions are `[start, end)`, meaning the region's
+  /// starting point is included in the interval, but the end point
+  /// is not. This is done in order to support fully-adjacent regions,
+  /// e.g. two function declarations which are alongside each other with
+  /// a `@warn` attribute on each will result in two warning control regions
+  /// where the first region ends at the exact same `AbsolutePosition` as
+  /// the start of the next region.
+  func contains(_ position: AbsolutePosition) -> Bool {
+    range.lowerBound <= position && position < range.upperBound
+  }
+
+  /// Check if this node's bounds contain another node
+  func contains(_ node: WarningControlRegionNode) -> Bool {
+    range.lowerBound <= node.range.lowerBound && node.range.upperBound <= range.upperBound
+  }
+
+  /// Check if this node's bounds match another node's
+  func hasSameBoundaries(with node: WarningControlRegionNode) -> Bool {
+    range.lowerBound == node.range.lowerBound && range.upperBound == node.range.upperBound
```

Comment:
```suggestion
    range == node.range
```

Which kind of makes you wonder if any of these `contains`-style methods are needed or if callers should just access `range` and call whatever they want on it.